sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1024,6 +1028,9 @@ object CombineUnions extends Rule[LogicalPlan] {
         case Distinct(Union(children, byName, allowMissingCol))
             if flattenDistinct && byName == topByName && allowMissingCol == topAllowMissingCol =>
           stack.pushAll(children.reverse)
+        case Deduplicate(_, Union(children, byName, allowMissingCol))
```

Comment:
How about we only handle distinct-like `Deduplicate`, where the `keys` == `output`

Reply:
Do you mean we don't need to optimize this case :
```
val df1 = Seq((1, 2, 3)).toDF("a", "b", "c")
val df2 = Seq((1, 2, 5)).toDF("a", "b", "c")
val df3 = Seq((2, 4, 3)).toDF("a", "b", "c")
val df4 = Seq((1, 4, 5)).toDF("a", "b", "c")
df1.union(df2).dropDuplicates(Seq("a")).union(df3).dropDuplicates("c").union(df4).dropDuplicates("b")
``` 
Only when the keys of Deduplicate is equal to the output of Union?

Reply:
Done